QEfficient/transformers/models/gemma/modeling_gemma.py

```diff
@@ -0,0 +1,572 @@
+        super(GemmaRotaryEmbedding, self).__init__()
+        self.dim = dim
+        self.max_position_embeddings = max_position_embeddings
+        self.base = base
+        inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2, dtype=torch.int64).float().to(device) / self.dim))
+        self.register_buffer("inv_freq", inv_freq, persistent=False)
+
+        # Build here to make `torch.jit.trace` work.
+        self._set_cos_sin_cache(
+            seq_len=max_position_embeddings, device=self.inv_freq.device, dtype=torch.get_default_dtype()
+        )
+
+    def _set_cos_sin_cache(self, seq_len, device, dtype):
+        self.max_seq_len_cached = seq_len
+        t = torch.arange(self.max_seq_len_cached, device=device, dtype=torch.int64).type_as(self.inv_freq)
+
+        freqs = torch.outer(t, self.inv_freq)
+
+        emb = torch.cat((freqs, freqs), dim=-1)
+        self.register_buffer("cos_cached", emb.cos().to(dtype), persistent=False)
+        self.register_buffer("sin_cached", emb.sin().to(dtype), persistent=False)
+
+    def forward(self, x, seq_len=None):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        if seq_len > self.max_seq_len_cached:
+            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)
+
+        return (
+            self.cos_cached[:seq_len].to(dtype=x.dtype),
+            self.sin_cached[:seq_len].to(dtype=x.dtype),
+        )
+
+
+def qeff_apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim=1):
+    """Applies Rotary Position Embedding to the query and key tensors.
+
+    Args:
+        q (`torch.Tensor`): The query tensor.
+        k (`torch.Tensor`): The key tensor.
+        cos (`torch.Tensor`): The cosine part of the rotary embedding.
+        sin (`torch.Tensor`): The sine part of the rotary embedding.
+        position_ids (`torch.Tensor`):
+            The position indices of the tokens corresponding to the query and key tensors. For example, this can be
+            used to pass offsetted position ids when working with a KV-cache.
+        unsqueeze_dim (`int`, *optional*, defaults to 1):
+            The 'unsqueeze_dim' argument specifies the dimension along which to unsqueeze cos[position_ids] and
+            sin[position_ids] so that they can be properly broadcasted to the dimensions of q and k. For example, note
+            that cos[position_ids] and sin[position_ids] have the shape [batch_size, seq_len, head_dim]. Then, if q and
+            k have the shape [batch_size, heads, seq_len, head_dim], then setting unsqueeze_dim=1 makes
+            cos[position_ids] and sin[position_ids] broadcastable to the shapes of q and k. Similarly, if q and k have
+            the shape [batch_size, seq_len, heads, head_dim], then set unsqueeze_dim=2.
+    Returns:
+        `tuple(torch.Tensor)` comprising of the query and key tensors rotated using the Rotary Position Embedding.
+    """
+    cos = cos[position_ids].unsqueeze(unsqueeze_dim)
```

Comment:
@quic-meet . Since you have mentioned you have referred llama rotary embedding. The current llama one has rope scaling as well, because higher PL, CL was causing incorrect outputs.

Reply:
Please run the perplexity script as well, and paste the results too in PR

Reply:
Please run for larger CL=1024 PL combination and get perplexity results. This way we can confirm if scaling might be needed or not.

Reply:
Sure @quic-akuruvil. Below are the results for torch and qpc

Torch:
```
Sample No:0 	 AVG_LOSS: 1.9312
2024-10-27 22:40:39,292 [INFO] E2E Sample Time: 10.5456s	 E2E TOKENS/S : 194.11
Sample No:1 	 AVG_LOSS: 2.5719
2024-10-27 22:40:55,867 [INFO] E2E Sample Time: 16.4811s	 E2E TOKENS/S : 124.20
Sample No:2 	 AVG_LOSS: 2.9851
2024-10-27 22:41:08,261 [INFO] E2E Sample Time: 12.2962s	 E2E TOKENS/S : 166.47
2024-10-27 22:41:08,371 [INFO] TORCH Perplexity: 12.1343
2024-10-27 22:41:08,371 [INFO] TORCH Loss: 2.4960
2024-10-27 22:41:08,371 [INFO] Total time for evaluation: 0.01 hrs

Loading Dataset: wikitext-2-raw-v1
Loading Model From: None
Model Type Mentioned: torch
Samples for Inference: 3
Batch Size: 1
Context Length: 2048
Dataset Prompt Length: 1
Dataset Stride: 1024
Overall Loss: 2.4960343837738037
Perplexity: 12.134278297424316
Total time for evaluation: 0.01342263917128245 hrs

*******************************************************
Torch Original Perplexity: 12.134278297424316
Target Perplexity for FP16 Precision: 12.146411895751953
Target Perplexity for MXFP6/MXINT8 Precision: 12.255620956420898
*******************************************************
```

QPC:
```
sample_no:0 	 avg_loss: 1.9440
2024-10-27 22:35:43,181 [INFO] e2e sample time: 99.3123s	 e2e tokens/s : 20.61
sample_no:1 	 avg_loss: 2.7717
2024-10-27 22:37:21,082 [INFO] e2e sample time: 97.9004s	 e2e tokens/s : 20.91
sample_no:2 	 avg_loss: 2.9682
2024-10-27 22:38:57,912 [INFO] e2e sample time: 96.8287s	 e2e tokens/s : 21.14
2024-10-27 22:38:57,985 [INFO] QPC Perplexity: 12.9524
2024-10-27 22:38:57,985 [INFO] QPC Loss: 2.5613
2024-10-27 22:38:57,986 [INFO] Total time for evaluation: 0.09 hrs

Loading Dataset: wikitext-2-raw-v1
Loading Model From: /prj/qct/aisyssol_scratch/users/meetdosh/.cache/qeff_cache/google/gemma-2b/qpc_16cores_1bs_2pl_2048cl_-1mos_1devices_fp16/qpcs/
Model Type Mentioned: qpc
Samples for Inference: 3
Batch Size: 1
Context Length: 2048
Dataset Prompt Length: 1
Dataset Stride: 1024
Overall Loss: 2.561280091603597
Perplexity: 12.95238694973654
Total time for evaluation: 0.08586144963900248 hrs
```